pkg/storage/mvcc.go

```diff
@@ -5332,6 +5332,144 @@ func CanGCEntireRange(
+		// reports how many keys were GC'd. The approach is identical to what point
+		// key GC does for consistency, but both places could be improved.
+		log.Eventf(ctx,
+			"done with GC evaluation for clear range of %d keys at %.2f keys/sec. Deleted %d entries",
+			countKeys, float64(countKeys)*1e9/float64(timeutil.Since(begin)), removedEntries)
+	}(timeutil.Now())
+
+	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+		KeyTypes:   IterKeyTypePointsAndRanges,
+	})
+	defer iter.Close()
+
+	iter.SeekGE(MVCCKey{Key: start})
+
+	var (
+		prevPointKey    MVCCKey
+		rangeTombstones MVCCRangeKeyStack
+		firstKey        = true
+	)
+
+	for ; ; iter.Next() {
+		if ok, err := iter.Valid(); err != nil {
+			return err
+		} else if !ok {
+			break
+		}
+
+		if iter.RangeKeyChanged() {
+			iter.RangeKeys().CloneInto(&rangeTombstones)
+			if hasPoint, _ := iter.HasPointAndRange(); !hasPoint {
+				continue
+			}
+		}
+
+		// Invariant: we're now positioned on a point key. The iterator can only
+		// be positioned on a bare range key when `RangeKeyChanged()` returns `true`.
+		countKeys++
+		unsafeKey := iter.UnsafeKey()
+		newKey := !prevPointKey.Key.Equal(unsafeKey.Key)
+		if newKey {
+			unsafeKey.CloneInto(&prevPointKey)
+			prevPointKey.Timestamp = hlc.Timestamp{}
+		}
+
+		// Skip keys that fall outside of range (only until we reach the first
+		// eligible key).
+		if firstKey && unsafeKey.Compare(MVCCKey{Key: start, Timestamp: startTimestamp}) < 0 {
+			prevPointKey.Timestamp = unsafeKey.Timestamp
+			continue
+		}
+		firstKey = false
+
+		if unsafeKey.Timestamp.IsEmpty() {
+			// Found intent
+			return errors.Errorf("attempt to GC intent %s using clear range",
+				unsafeKey.String())
+		}
+		if gcThreshold.Less(unsafeKey.Timestamp) {
+			// Current version is above GC threshold so it is not safe to clear.
+			return errors.Errorf("attempt to GC data %s above threshold %s with clear range",
+				unsafeKey.String(), gcThreshold.String())
+		}
+
+		valueLen, isTombstone, err := iter.MVCCValueLenAndIsTombstone()
+		if err != nil {
+			return err
+		}
+
+		// Find timestamp covering current key.
+		coveredBy := prevPointKey.Timestamp
+		if rangeKeyCover, ok := rangeTombstones.FirstAtOrAbove(unsafeKey.Timestamp); ok {
+			// If there's a range between current value and value above
+			// use that timestamp.
+			if coveredBy.IsEmpty() || rangeKeyCover.Timestamp.Less(coveredBy) {
+				coveredBy = rangeKeyCover.Timestamp
+			}
+		}
+		isCovered := !coveredBy.IsEmpty() && coveredBy.LessEq(gcThreshold)
+
+		validTill := coveredBy
+		if isTombstone {
+			validTill = unsafeKey.Timestamp
+		}
+
+		if !isCovered && !isTombstone {
+			// Current version is below threshold, but preceding one is not so it is
+			// visible and can't be cleared.
+			return errors.Errorf("attempt to GC data %s still visible at GC threshold %s with clear range",
+				unsafeKey.String(), gcThreshold.String())
```

Comment:
nit: we don't need `.String()` here and elsewhere, `%s` will automatically call it.

Reply:
Well not exactly. Errorf uses Formatter first and our formatter implementation always prints timestamp but String() respects unversioned keys and omits timestamp. This is annoying a bit. I think fmt.Printf uses String if it is available in that case. But changing formatter will blow up all tests. :(